### PR TITLE
Added ClusterStaticEntry support

### LINKED ISSUE #149: Proposal: ClusterStaticEntry CRD for registering workloads that live outside the cluster
It has become apparent that the spire-controller-manager needs to manage registration entries for workloads that are not running inside the cluster (see #76)

To satisfy this need we propose the introduction of the `ClusterStaticEntry` CRD.

During reconciliation, the spire-controller-manager will formulate the set of entries that it reconciles against SPIRE server based on:
- entries for pods that are selected by each `ClusterSPIFFEID` resource  (current behavior)
- entries rendered from `ClusterStaticEntry` resources.

The `ClusterStaticEntry` CRD would have the following fields:

- 	SpiffeID string
- 	ParentID string
- 	Selectors string
- 	FederatesWith []string
- 	X509SVIDTTL metav1.Duration
- 	JWTSVIDTTL metav1.Duration
- 	DnsNames []string
- 	Hint string
- 	Admin bool
- 	Downstream bool

### PR DESCRIPTION
**Pull request by azdagron**
Resolves #149

### CODE REVIEW

File: config/crd/bases/spire.spiffe.io_clusterstaticentries.yaml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+              selectors:
+                items:
+                  type: string
+                type: array
+              spiffeID:
+                type: string
+              x509SVIDTTL:
+                type: string
+            required:
+            - dnsNames
```
Comment (MarcosDY) [inline config/crd/bases/spire.spiffe.io_clusterstaticentries.yaml:66]: is it intntional to make dns, federatesWith, hint and ttls  required?

File: README.md [hunk 2]
```diff
@@ -64,6 +74,27 @@ The [demo](demo) includes [sample configuration](demo/config/cluster1) for
 deploying the SPIRE Controller Manager, SPIRE, and the SPIFFE CSI driver,
 including requisite RBAC and Webhook configuration.
 
+## Compatibility
+
+The SPIRE APIs used by the SPIRE Controller Manager are generally stable and
+supported since at least SPIRE v1.0. However, the API has gained support for
+additional entry fields beyond what was supported in SPIRE v1.0. Notably, these
+include both the `jwt_svid_ttl` and the `hint` fields. The ClusterStaticEntry
```
[the remark this answers is not in the document]
Comment (azdagron) [reply README.md:82]: jwt_svid_ttl makes sense but hint is harder to reason about. I'd probably side with waiting until somebody has a use case for it before adding it to the ClusterSPIFFEID CRD.

File: api/v1alpha1/clusterstaticentry_types.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2021 SPIRE Authors.
```
[the remark this answers is not in the document]
Comment (azdagron) [reply api/v1alpha1/clusterstaticentry_types.go:2]: I think we'd only bump the year if there were significant updates. I'll change the year for this CRD.

File: config/crd/bases/spire.spiffe.io_clusterstaticentries.yaml (hunk 1 above)
Comment (azdagron) [reply config/crd/bases/spire.spiffe.io_clusterstaticentries.yaml:66]: Good catch. I've updated the json tags and regenerated manifests.

File: config/samples/spire_v1alpha1_clusterstaticentry.yaml
Comment (ugoenyioha) [inline config/samples/spire_v1alpha1_clusterstaticentry.yaml:1]: we should complete the sample here no? else what would users have to anchor on?

**Review by MarcosDY — APPROVED** [review]
LGTM!!!

**Comment by matthewdevenny** [comment]
Nice! This is definitely needed - what about the `node` setting? If I am using the `boxboat/spire-tpm-plugin` for node tpm attestation I would also need to be able to set the `node` flag correct?

**Comment by azdagron** [comment]
The -node flag doesn't actually have a first class field on the entry. Instead it sets the parent ID to the SPIFFE ID of the spire server (i.e., `spiffe://TRUSTDOMAIN/spire/server`), which implies that the entry is a node alias.

**Comment by keeganwitt** [comment]
We should probably also create a migration guide.

**Comment by kfox1111** [comment]
Yeah.... upgrading the chart will likely have some issues with the crds.

**Pull request merged**